### PR TITLE
Offline-First TWA doc

### LINKED ISSUE #693: Offline-First TWAs [4/30/2021]
* [Draft](https://docs.google.com/document/d/19j7f9Q75vvMyH8vBQQcdV0kctPoTtDs6fcbfomq20xw/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by demianrenzulli**
Fixes #693

Offline-First TWA doc. Companion article for [demo](https://github.com/GoogleChrome/android-browser-helper/tree/main/demos/twa-offline-first) in [android-browser-helper repo](https://github.com/GoogleChrome/android-browser-helper).

cc // @jpmedley @kaycebasques @andreban

### CODE REVIEW

**Comment by jpmedley** [comment]
@demianrenzulli Are all technical and content reviews finished? Copy edits are best done after text has stablized.

**Comment by demianrenzulli** [comment]
Hi @jpmedley! 

So far, I received your comments and a review by @andreban. The [demo](https://github.com/GoogleChrome/android-browser-helper/tree/main/demos/twa-offline-first) this article is based on was reviewed by @PEConn.

From the thread with @kaycebasques, it doesn't seem we need any additional reviewers.

**Comment by jpmedley** [comment]
Kayce was probably reviewing for content appropriateness and structure. There's always a final review for typos and grammar mistakes.

@kaycebasques What type of review did you do?

**Comment by demianrenzulli** [comment]
I don't think @kaycebasques did a review yet, he just connected us with you for next steps. If Kayce can take another look, that would be great! I just don't want to give you guys so much work.

Thanks for the help!

**Comment by jpmedley** [comment]
@demianrenzulli Before we edit this, can you please break the lines at 80 characters. You'll notice that all of our article sources are broken. It has certain advantages during the editing process.

Thanks.

**Comment by demianrenzulli** [comment]
@jpmedley I just ran [rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) to break each line at the 80 chars. I hope it helps!

**Comment by jpmedley** [comment]
Thank you very much.
Joe Medley | Technical Writer, Chrome DevRel | ***@***.*** |
 816-678-7195
*If an API's not documented it doesn't exist.*

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+{% Aside %} A prerequisite of this guide is to have a Trusted Web Activity app
+running. Follow the steps in the [Integration
+Guide](https://developer.chrome.com/docs/android/trusted-web-activity/integration-guide/)
+to create a basic Trusted Web Activity project. {% endAside%}
+
+## Create a custom LauncherActivity
+
+The first step is to create a custom launcher activity. This is the `Activity`
+that will contain the offline screen to show if the first time the user opens
+the app there's no connectivity.
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:41]:
```suggestion
The first step is to create a custom launcher activity. This `Activity`
that will contain the offline screen to show if there's no connectivity
the first time a user opens the app.
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        <category android:name="android.intent.category.DEFAULT" />
+        <category android:name="android.intent.category.BROWSABLE" />
+        <!-- Edit android:host to handle links to the target URL -->
+        <data android:host="airhorner.com" android:scheme="https" />
+    </intent-filter>
+</activity>
+```
+
+The previous code registers `OfflineFirstTWALauncherActivity` as a launcher
+Activity and defines [https://airhorner.com](https://airhorner.com) as the URL
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:76]:
```suggestion
activity and defines [https://airhorner.com](https://airhorner.com) as the URL
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+to open when the TWA launches. 
+
+## Handle offline scenarios
+
+First, inside the Activity, override the method `shouldLaunchImmediately()` and
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:81]:
```suggestion
First, inside the Activity, override the `shouldLaunchImmediately()` method and
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+make it return `false`, so that the Trusted Web Activity won't launch
+immediately, and you can add some extra checks before the initial launch:
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:83]:
```suggestion
immediately. You can also add extra checks before the initial launch:
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        firstTimeLaunchTwa();
+    } else {
+        renderOfflineFallback();
+    }
+}
+```
+
+The previous code handles three scenarios:
+
+- If the TWA has launched previously, this means that the service worker has
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:125]:
```suggestion
- If the TWA has launched previously, the service worker has
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  been registered, and the PWA will be able to respond offline. In that case,
+  call the method `launchTwa()` defined in the parent class to launch the
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:127]:
```suggestion
  call `launchTwa()`, defined in the parent class, to launch the
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+- If the TWA hasn't launched previously and the user is online, launch the
+  Trusted Web Activity for the first time using the `firstTimeLaunchTwa()`
+  method that you'll implement later.
+- If the TWA hasn't already launched and the user is offline, render the native
+  offline fallback screen.
+
+## Implement helper methods
+
+The final step is to implement the helper methods called by the previous code.
+Here's how the code to check of offline state `isOnline()` looks like:
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:138]:
```suggestion
Here's the code for checking offline state `isOnline()`:
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+```java
+private boolean isOnline() {
+    ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
+    NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
+    return activeNetworkInfo != null && activeNetworkInfo.isConnected();
+}
+```
+
+Next, implement `hasTwaLaunchedSuccessfully()` used to check if the TWA has
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:148]:
```suggestion
Next, implement `hasTwaLaunchedSuccessfully()`, which checks if the TWA has
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+```java
+private boolean hasTwaLaunchedSuccessfully() {
+    // Return `true` if the preference "twa_launched_successfully" has already been set.
+    SharedPreferences sharedPref = getSharedPreferences(getString(R.string.twa_offline_first_preferences_file_key), Context.MODE_PRIVATE);
+    return sharedPref.getBoolean(getString(R.string.twa_launched_successfully), false);
+}
+```
+
+The previous code calls the `launchTWA()` from the parent class, and saves the
+`twa_launched_successfully` flag in shared preferences, to indicate that the TWA
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:160]:
```suggestion
`twa_launched_successfully` flag in shared preferences, to indicate that the TWA
```
```suggestion
`twa_launched_successfully` flag in shared preferences. This indeicates that the TWA
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+has launched successfully, at least once.
+
+The remaining helper method: `renderOfflineFallback()` renders an Android
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:163]:
```suggestion
The remaining helper method, `renderOfflineFallback()` renders an Android
```

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            // Check connection status. If online, launch the Trusted Web Activity for the first time.
+            if (isOnline()) firstTimeLaunchTwa();
+        }
+    });
+}
+```
+
+Fort his demo, we have defined the `activity_offline_first_twa` layout, which
+contains a button to retry, which will, in time, to execute
+`firstTimeLaunchTwa()` after checking connection:
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:182]:
```suggestion
`firstTimeLaunchTwa()` after checking the connection.
```

Comment (jpmedley) [reply site/en/docs/android/trusted-web-activity/offline-first/index.md:182]: Colons aren't used to introduce images. If you feel that the image needs to be referenced in text, add something like, 'see the image below'.

File: site/en/docs/android/trusted-web-activity/offline-first/index.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+- The first time a user launches a Progressive Web App (PWA) via Trusted Web
+  Activity, the service worker won't yet be available.
+- To avoid showing the standard offline screen if the user has no connectivity,
+  you can detect the offline condition and show a fallback offline screen
+  instead.
+- In this guide you learned how to implement that strategy. If you are
+  interested in checking the code we used throughout this guide, the complete
+  solution can be found
+  [here](https://github.com/GoogleChrome/android-browser-helper/tree/main/demos/twa-offline-first).
```
Comment (jpmedley) [inline site/en/docs/android/trusted-web-activity/offline-first/index.md:197]:
```suggestion
  interested in checking the code we used throughout this guide, you can find
  the complete solution in the [Offline First TWA
  Demo](https://github.com/GoogleChrome/android-browser-helper/tree/main/demos/twa-offline-first).
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

**Comment by demianrenzulli** [comment]
Hi @jpmedley! Thanks again for leaving your suggestions. I just went through all of them and committed the changes.

Please, let me know if you need anything else from me.

**Review by jpmedley — APPROVED** [review]

**Comment by jpmedley** [comment]
@demianrenzulli DevRel members merge their own articles after approval. I don't believe you have the authority. Let me know if you need me to merge this.

**Comment by demianrenzulli** [comment]
@jpmedley I remember Kayce asked to merge some posts myself in Web.dev, but I never wrote something for developer.chrome.com, so I would appreciate if you can marge on my behalf.

**Pull request merged**